llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -4401,6 +4401,10 @@ class TargetLowering : public TargetLoweringBase {
   /// Return true if the target supports ptrauth operand bundles.
   virtual bool supportPtrAuthBundles() const { return false; }
 
+  /// Return true if the target supports FCOPYSIGN nodes with operands of
+  /// mismatched types.
+  virtual bool supportMismatchCopysign() const { return true; }
```

Comment:
-1 on adding this. It's not hard to match the mismatched cases. Also, this should be expressible in terms of the legality of the type combination

Reply:
Fair enough. To be honest, I think it might be best to change the type of FCOPYSIGN since it seems like most targets do not benefit from the mismatched types and must instead add additional logic to re-insert the proper extension/truncation. That being said for the purposes of this change I've removed this API and instead added custom lowering to a node with matching types.

Reply:
This is mostly yet another instance of SDAG being bad with legality rules that aren't expressive enough. globalisel fixes this by having 2 type parameters for copysign